File: engine/bridge.py
"""JSON bridge — the contract crossing the Pyodide boundary (HANDOFF_PLAN §3).

Pure JSON text in / JSON text out, stateful via a branded string handle. The
solve-once state lives in Python; JS keeps the handle and asks for a *whole time
grid* at once (the slider then scrubs a cursor over the returned arrays
client-side — no per-tick re-solve, no PyProxy lifetime to manage).

Errors cross as a loud structured ``{"ok": false, "error": {type, message}}`` —
never a swallowed exception, never a fabricated fallback number (CLAUDE.md
no-silent-errors). Success responses carry ``"ok": true``.
"""

from __future__ import annotations

import json
import re
import secrets
import traceback

import radioactivedecay as rd

from engine import attenuation, buildup
from engine.attenuation import AttenuationError
from engine.beta_dose import BetaDoseError, BetaSkinDoseModel
from engine.buildup import BuildupError
from engine.chain import build_dag
from engine.conversion import ConversionError
from engine.decay_heat import DecayHeatError, DecayHeatModel
from engine.dose import SCORING_FLOOR_MEV, DoseError, GammaDoseModel
from engine.emissions import EmissionsError
from engine.inventory import EngineError, SolvedInventory
from engine.neutron_dose import NeutronDoseError, NeutronDoseModel
from engine.neutron_source import NeutronSourceError
from engine.photon_interp import OffGridError

#: Expected, structured domain errors — surfaced loudly but without a traceback (the
#: message is the contract). An *un*expected exception still carries its traceback.
_EXPECTED_ERRORS = (
    EngineError,
    DoseError,
    DecayHeatError,
    BetaDoseError,
    NeutronDoseError,
    NeutronSourceError,
    OffGridError,
    AttenuationError,
    BuildupError,
    ConversionError,
    EmissionsError,
)

_REGISTRY: dict[str, SolvedInventory] = {}


def _ok(payload: dict) -> str:
    return json.dumps({"ok": True, **payload})


def _err(exc: Exception) -> str:
    return json.dumps(
        {
            "ok": False,
            "error": {
                "type": type(exc).__name__,
                "message": str(exc),
                # full traceback only for unexpected (non-domain) failures
                "traceback": None if isinstance(exc, _EXPECTED_ERRORS) else traceback.format_exc(),
            },
        }
    )


def _solve_obj(nuclides: dict, unit: str = "Bq", precision: str = "double") -> SolvedInventory:
    """Internal: build a SolvedInventory (used by ``solve`` and by tests)."""
    return SolvedInventory.from_spec(nuclides, unit, precision)


_NUCLIDE_RE = re.compile(r"^([A-Za-z]+)-(\d+)([a-z]*)$")


def _nuclide_sort_key(name: str) -> tuple:
    """Natural (element, mass, isomeric-state) order so the picker lists Co-58
    before Co-60 before Cs-137 — not naive string order (where 'Co-60' < 'Co-58').
    Names that don't parse fall back to lexical order, kept (never dropped)."""
    m = _NUCLIDE_RE.match(name)
    if not m:
        return (1, name, 0, "")
    element, mass, state = m.groups()
    return (0, element, int(mass), state)


def nuclides() -> str:
    """``-> {ok, nuclides: [...]}`` — every nuclide the engine can solve (rd's full
    dataset), the add-by-name source for the M6b inventory panel. Names only, sorted
    and de-duplicated; emission availability is a *dose*-time concern (M6f), not here,
    so an emission-less but solvable nuclide is intentionally listed."""
    try:
        names = sorted({str(n) for n in rd.DEFAULTDATA.nuclides}, key=_nuclide_sort_key)
        return _ok({"nuclides": names})
    except Exception as exc:  # noqa: BLE001 - surfaced loudly as structured error
        return _err(exc)


def materials() -> str:
    """``-> {ok, materials: [{id, has_buildup, density_g_cm3}]}`` — the M6g shield-builder
    material list, the no-drift source for the §9 picker. Every material with a bundled
    attenuation file, tagged with ``has_buildup`` (an ANS-6.4.3 G-P buildup file exists).

    ``has_buildup`` is the **γ-shield gate**: ``GammaDoseModel`` *raises* for a shield
    material with no buildup (dose.py — a shield without scatter buildup is a data hole,
    not a transparent medium), so the UI filters the γ picker to ``has_buildup`` rather
    than letting a selection error out the whole γ panel (§11). Low-Z hydrogenous materials
    (PMMA, polyethylene) are listed with ``has_buildup=false`` — surfaced, not hidden — and
    become selectable for β/neutron once those consumers land (M7)."""
    try:
        out = [
            {
                "id": m,
                "has_buildup": buildup.has_material(m),
                "density_g_cm3": attenuation.density(m),
            }
            for m in sorted(attenuation.available_materials())
        ]
        return _ok({"materials": out})
    except Exception as exc:  # noqa: BLE001 - surfaced loudly as structured error
        return _err(exc)


def _get(handle: str) -> SolvedInventory:
    solved = _REGISTRY.get(handle)
    if solved is None:
        raise EngineError(f"unknown or released handle {handle!r}")
    return solved


def solve(spec_json: str) -> str:
    """Two accepted forms (dispatched on the presence of ``entries``):

    * single-unit: ``{"nuclides": {name: value}, "unit": "Bq", "precision": "double"}``
    * per-entry units (the §9 inventory panel): ``{"entries": [{"name","quantity","unit"}],
      "precision": "double"}`` — each entry may use a different unit, summed in atoms.

    -> ``{ok, handle, nuclides, half_lives_s, time_range_s, hp_recommended, ...}``."""
    try:
        spec = json.loads(spec_json)
        precision = spec.get("precision", "double")
        if "entries" in spec:
            solved = SolvedInventory.from_entries(spec["entries"], precision)
        else:
            solved = _solve_obj(spec["nuclides"], spec.get("unit", "Bq"), precision)
        handle = "inv_" + secrets.token_hex(8)
        _REGISTRY[handle] = solved
        return _ok({"handle": handle, **solved.metadata()})
    except Exception as exc:  # noqa: BLE001 - surfaced loudly as structured error
        return _err(exc)


def evaluate(handle: str, request_json: str) -> str:
    """``{"times_s": [...], "axis": "activity", "unit": "Bq"}`` ->
    ``{ok, axis, unit, times_s, nuclides, series, peak_atoms, floor_atoms, ...}``."""
    try:
        req = json.loads(request_json)
        solved = _get(handle)
        return _ok(solved.evaluate(req["times_s"], req.get("axis", "activity"), req.get("unit")))
    except Exception as exc:  # noqa: BLE001
        return _err(exc)


def chain(handle: str) -> str:
    """-> ``{ok, nodes, edges}`` (the decay-chain DAG)."""
    try:
        return _ok(build_dag(_get(handle)))
    except Exception as exc:  # noqa: BLE001
        return _err(exc)


def dose(handle: str, request_json: str) -> str:
    """``{"times_s":[...], "quantity":"ambient_H10", "distance_m":1.0,
        "shield":["lead",2.0] | null, "medium":"air", "geometry":null}`` ->
    ``{ok, quantity, si_unit, per, times_s, distance_m, rate_si, scoring_floor_MeV,
       warnings}``.

    The §6 gamma dose-rate over a whole time grid, the §3 "solve once, evaluate many" way:
    one Bateman solve (the handle), one ``GammaDoseModel`` (the per-nuclide dose coefficients
    ``C_n``), one matvec against the activity series. The slider scrubs the returned array
    client-side — no per-tick re-solve, no per-tick line summation."""
    try:
        req = json.loads(request_json)
        solved = _get(handle)
        activities = solved.evaluate(req["times_s"], axis="activity", unit="Bq")
        shield = req.get("shield")
        model = GammaDoseModel(
            solved.names,
            req.get("quantity", "ambient_H10"),
            medium=req.get("medium", "air"),
            shield=tuple(shield) if shield else None,
            geometry=req.get("geometry"),
        )
        return _ok(model.dose_rate_series(activities, float(req["distance_m"])))
    except Exception as exc:  # noqa: BLE001 - surfaced loudly as structured error
        return _err(exc)


def dose_lines(handle: str, request_json: str) -> str:
    """``{"quantity":"ambient_H10", "geometry":null, "shield":["lead",2.0]|null,
        "medium":"air"}`` -> ``{ok, quantity, si_unit, lines:[{nuclide, E_MeV, yield,
        origin, coeff_si}], warnings, scoring_floor_MeV}``.

    The §9 per-line γ breakdown ("the gamma slice expands to a per-line table"). DISTANCE-
    and TIME-FREE on purpose: it returns only the per-line per-decay coefficients ``coeff_si``
    (Sv·m²/decay for H*(10)/effective; J·m²/kg for air_kerma). The client applies the
    geometric factor ``1/4πd²`` and the parent's activity ``A_n(t)`` at the cursor, so the
    table is live on scrub/distance with no re-fetch — the §3 "solve once, evaluate many"
    discipline the dose series already follows. ``coeff_si`` is the same per-line quantity
    that sums to ``dose()``'s ``C_n``, so ``Σ(coeff_si·A_n)/4πd²`` reconciles EXACTLY with
    the ``dose()`` total (one coefficient-assembly path, no second code branch to drift)."""
    try:
        req = json.loads(request_json)
        solved = _get(handle)
        shield = req.get("shield")
        model = GammaDoseModel(
            solved.names,
            req.get("quantity", "ambient_H10"),
            medium=req.get("medium", "air"),
            shield=tuple(shield) if shield else None,
            geometry=req.get("geometry"),
        )
        return _ok(
            {
                "quantity": model.quantity,
                "si_unit": model.si_unit,
                "lines": model.per_line_rows(),
                "warnings": list(model.warnings),
                "scoring_floor_MeV": SCORING_FLOOR_MEV,
            }
        )
    except Exception as exc:  # noqa: BLE001 - surfaced loudly as structured error
        return _err(exc)


def dose_thickness(handle: str, request_json: str) -> str:
    """``{"material":"lead", "thicknesses_cm":[0,0.5,1,...], "quantity":"ambient_H10",
        "geometry":null, "medium":"air"}`` -> ``{ok, quantity, si_unit, material,
        thicknesses_cm, coeff_by_nuclide:{nuclide:[C_n(x0), C_n(x1), ...]}, warnings,
        scoring_floor_MeV}``.

    The §9 **dose-vs-thickness** sweep, Design-A (M6f-2 #10): DISTANCE- and TIME-free
    per-nuclide γ dose coefficients ``C_n(x)`` for a thickness grid through one ``material``.
    The shield transmission ``B(E,μx)·exp(−μx)`` is nonlinear and per-line, so — unlike the
    γ dose-vs-*distance* band (exact inverse-square, reconstructed client-side) — it MUST be
    evaluated by the engine. The client folds ``1/4πd²`` and the parent activity ``A_n(t)``
    at the cursor, so the curve is live on scrub/distance with no re-fetch (the §3 discipline
    the dose series already follows).

    ``x == 0`` is evaluated with ``shield=None`` so the curve's zero point is the EXACT
    unshielded baseline. Each ``C_n(x)`` is the same per-nuclide coefficient ``dose()`` folds
    for ``shield=(material, x)`` (one ``GammaDoseModel`` assembly path), so at any grid
    thickness ``Σ C_n(x)·A_n / 4πd²`` reconciles EXACTLY with the breakdown bar — no second
    code branch to drift. Below-floor lines are logged in ``warnings`` (thickness-independent,
    captured once); a material without ANS-6.4.3 buildup raises loudly (§11)."""
    try:
        req = json.loads(request_json)
        solved = _get(handle)
        material = str(req["material"])
        quantity = req.get("quantity", "ambient_H10")
        geometry = req.get("geometry")
        medium = req.get("medium", "air")
        thicknesses = [float(x) for x in req["thicknesses_cm"]]

        coeff_by_nuclide: dict[str, list[float]] = {n: [] for n in solved.names}
        warnings: list[dict] = []
        si_unit = None
        for i, x in enumerate(thicknesses):
            if x < 0.0:
                raise DoseError(f"shield thickness must be >= 0 cm; got {x}")
            model = GammaDoseModel(
                solved.names,
                quantity,
                medium=medium,
                shield=(material, x) if x > 0.0 else None,
                geometry=geometry,
            )
            si_unit = model.si_unit
            for n in solved.names:
                coeff_by_nuclide[n].append(model.coeff_si[n])
            # Below-floor skips are set by the line *energies* / quantity, not the shield, so
            # they are identical at every thickness — capture them once (the x=0 model).
            if i == 0:
                warnings = list(model.warnings)
        return _ok(
            {
                "quantity": quantity,
                "si_unit": si_unit,
                "material": material,
                "thicknesses_cm": thicknesses,
                "coeff_by_nuclide": coeff_by_nuclide,
                "warnings": warnings,
                "scoring_floor_MeV": SCORING_FLOOR_MEV,
            }
        )
    except Exception as exc:  # noqa: BLE001 - surfaced loudly as structured error
        return _err(exc)


def beta_dose(handle: str, request_json: str) -> str:
    """``{"times_s":[...], "distance_m":0.0, "shield":["pmma",1.0] | null,
        "medium":"tissue_soft", "avg_area_cm2":1.0, "include_bremsstrahlung":true,
        "brems_quantity":"ambient_H10", "geometry":null}`` ->
    ``{ok, quantity:"beta_skin", si_unit:"Gy", per, times_s, distance_m,
       scoring_depth_mg_cm2, rate_si, warnings, bremsstrahlung: {γ-dose series} | null}``.

    The §6.2 external **beta skin dose** (Hp(0.07), 7 mg/cm²) over a time grid, the §3
    "solve once, evaluate many" way: one solve, fixed per-nuclide C_n^β(d), one matvec. When
    a shield is present, the secondary **bremsstrahlung** photon dose is scored through the
    γ engine (a *different* quantity — penetrating photons, not skin dose) and returned
    alongside, so the UI can show "more lead can increase dose". Brems needs a finite
    distance (point-source γ is singular at 0)."""
    try:
        req = json.loads(request_json)
        solved = _get(handle)
        activities = solved.evaluate(req["times_s"], axis="activity", unit="Bq")
        shield = req.get("shield")
        model = BetaSkinDoseModel(
            solved.names,
            medium=req.get("medium", "tissue_soft"),
            shield=tuple(shield) if shield else None,
            avg_area_cm2=float(req.get("avg_area_cm2", 1.0)),
        )
        distance_m = float(req.get("distance_m", 0.0))
        out = model.dose_rate_series(activities, distance_m)
        out["bremsstrahlung"] = None
        if shield and req.get("include_bremsstrahlung", True):
            override = model.bremsstrahlung_override()
            if distance_m > 0.0 and any(override.values()):
                gm = GammaDoseModel(
                    solved.names,
                    req.get("brems_quantity", "ambient_H10"),
                    shield=None,  # beta-stopping shield is photon-thin (documented, §11)
                    geometry=req.get("geometry"),
                    photon_override=override,
                )
                out["bremsstrahlung"] = gm.dose_rate_series(activities, distance_m)
        return _ok(out)
    except Exception as exc:  # noqa: BLE001 - surfaced loudly as structured error
        return _err(exc)


def neutron_dose(handle: str, request_json: str) -> str:
    """``{"times_s":[...], "source":"Cf-252", "quantity":"ambient_H10", "distance_m":1.0,
        "geometry":null, "include_source_gamma":true}`` ->
    ``{ok, quantity, si_unit:"Sv", per, times_s, distance_m, source, parent,
       neutrons_per_decay, spectrum_avg_coeff_pSv_cm2, rate_si, warnings,
       source_gamma:{γ-dose series}|null}``.

    The §6.3 external **neutron dose** for a **prebuilt source** over a time grid, the §3
    "solve once, evaluate many" way: one Bateman solve (the handle), one ``NeutronDoseModel``
    (the fixed per-decay coefficient ``neutrons_per_decay·h̄``), one matvec against the
    parent's activity series. The ``"source"`` key is the §6.3 **gray-out gate** — the UI
    only calls this for a prebuilt neutron source, never for a user-defined inventory; the
    source's parent nuclide must be present in the handle's inventory (a missing parent is a
    loud error, never a silent zero).

    When the source carries reaction-correlated γ (e.g. AmBe 4.438 MeV — NOT in the ICRP-107
    decay lines), it is scored through the γ engine via ``photon_override`` in the **same
    quantity/geometry** and returned alongside, so the breakdown can show the photon
    contribution. (Cf-252 prompt-fission γ is unmodeled in v1 — §11 — so this is null there.)
    """
    try:
        req = json.loads(request_json)
        solved = _get(handle)
        activities = solved.evaluate(req["times_s"], axis="activity", unit="Bq")
        quantity = req.get("quantity", "ambient_H10")
        geometry = req.get("geometry")
        model = NeutronDoseModel(req["source"], quantity, geometry=geometry)
        distance_m = float(req["distance_m"])
        out = model.dose_rate_series(activities, distance_m)
        out["source_gamma"] = None
        if req.get("include_source_gamma", True):
            override = model.source_gamma_override()
            if any(override.values()):
                # Source-correlated γ scored in the SAME quantity/geometry as the neutron
                # dose, so the two are directly comparable in the breakdown.
                gm = GammaDoseModel(
                    [model.parent], quantity, geometry=geometry, photon_override=override
                )
                out["source_gamma"] = gm.dose_rate_series(activities, distance_m)
        return _ok(out)
    except Exception as exc:  # noqa: BLE001 - surfaced loudly as structured error
        return _err(exc)


def decay_heat(handle: str, request_json: str) -> str:
    """``{"times_s": [...]}`` -> ``{ok, quantity:"decay_heat", si_unit:"W", times_s,
        total_W, by_nuclide_W, coeff_W_per_Bq, E_rec_MeV, channels_MeV, definition}``.

    The §5 **decay heat** (thermal power, W) over a time grid, the §3 "solve once,
    evaluate many" way: one Bateman solve (the handle), one ``DecayHeatModel`` (the fixed
    per-nuclide recoverable-energy coefficient W/Bq, assembled once from the same bundled
    ICRP-107 emission spectra the γ/β engines read — no new dataset), one matvec against
    the activity series. DISTANCE- and quantity-free (heat is total locally-deposited power,
    not a point-field quantity). ``total_W`` is Σ over nuclides; ``by_nuclide_W`` is the
    §5/§9 dominant-contributor breakdown. ``definition`` carries the honesty-register
    statement of exactly what is summed (bulk recoverable energy, antineutrino-excluded)."""
    try:
        req = json.loads(request_json)
        solved = _get(handle)
        activities = solved.evaluate(req["times_s"], axis="activity", unit="Bq")
        model = DecayHeatModel(solved.names)
        return _ok(model.heat_series(activities))
    except Exception as exc:  # noqa: BLE001 - surfaced loudly as structured error
        return _err(exc)


def registry_size() -> str:
    """``-> {ok, size}`` — number of live solved inventories. The handle-leak canary
    for invariant #2 (solve-once / release-old): the UI keeps at most one live handle,
    so this stays at 1 across re-solves and 0 when the inventory is cleared."""
    try:
        return _ok({"size": len(_REGISTRY)})
    except Exception as exc:  # noqa: BLE001
        return _err(exc)


def release(handle: str) -> str:
    """Free a solved inventory. Idempotent; reports whether it existed."""
    try:
        existed = _REGISTRY.pop(handle, None) is not None
        return _ok({"handle": handle, "existed": existed})
    except Exception as exc:  # noqa: BLE001
        return _err(exc)

File: web/src/lib/bridge.ts
// Typed client for engine/bridge.py — the JSON contract crossing the Pyodide
// boundary (HANDOFF_PLAN §3). Pure JSON text in / JSON text out, stateful via a
// branded `Handle`. Every bridge response is the discriminated union
//   { ok: true, ... } | { ok: false, error: BridgeError }
// so callers must narrow on `.ok` before reading a payload — the no-silent-error
// contract (CLAUDE.md, §11) made structural: a failure can't be mistaken for data.

/** Opaque inventory handle minted by `solve`; never construct one by hand. */
export type Handle = string & { readonly __handle: unique symbol };

export interface BridgeError {
  type: string;
  message: string;
  /** Present only for unexpected (non-domain) failures. */
  traceback: string | null;
}

type Fail = { ok: false; error: BridgeError };
type Result<T> = (T & { ok: true }) | Fail;

// --- requests ----------------------------------------------------------------

/** One isotope line for the per-entry-units solve form (§9 inventory panel). */
export interface SolveEntry {
  name: string;
  quantity: number;
  /** An rd input unit: Bq, Ci, atoms, g, kg, mg, ug, ... */
  unit: string;
}

/**
 * Two accepted forms (the bridge dispatches on `entries`):
 *  - single-unit:   `{ nuclides, unit, precision }`
 *  - per-entry units: `{ entries: [{name,quantity,unit}], precision }`
 * The inventory panel always uses the entries form; selfcheck/tests use the older one.
 */
export type SolveSpec =
  | { entries: SolveEntry[]; precision?: "double" | "hp" }
  | {
      nuclides: Record<string, number>;
      /** Bq (default), Ci, atoms, g, kg, ug, ... — see inventory.from_spec. */
      unit?: string;
      precision?: "double" | "hp";
    };

export type DoseAxis = "atoms" | "mass" | "activity";

export interface EvaluateRequest {
  times_s: number[];
  axis?: DoseAxis;
  unit?: string | null;
}

/** A shield as the engine's ordered layer list (source-side → detector-side; the last
 *  layer is detector-adjacent). A single layer is `[["lead", 2.0]]`. Always the list
 *  form on the wire — the Python `_normalize_shield` reduces n=1 to the legacy path. */
export type ShieldSpec = [string, number][];

export interface DoseRequest {
  times_s: number[];
  quantity?: string; // "air_kerma" | "ambient_H10" | "effective"
  distance_m: number;
  shield?: ShieldSpec | null;
  medium?: string;
  geometry?: string | null;
}

/** Distance/time-free per-nuclide γ coefficients over a thickness grid (M6g dose-vs-
 *  thickness, Design-A). Two forms (M8): single-layer `{material}` (legacy; `x=0` is the
 *  exact unshielded baseline) or multi-layer `{layers, sweep_index}` — sweep that layer's
 *  thickness with the OTHERS held (then `x=0` is the rest-of-stack, not unshielded). The
 *  client folds `1/4πd²` and the cursor activity at every grid point (zero re-fetch). */
export interface DoseThicknessRequest {
  material?: string;
  layers?: ShieldSpec;
  sweep_index?: number;
  thicknesses_cm: number[];
  quantity?: string;
  geometry?: string | null;
  medium?: string;
}

// --- responses (only the fields the app reads are typed; the bridge may add more) ---

export interface SolveOk {
  handle: Handle;
  nuclides: string[];
  half_lives_s: Record<string, number>;
  time_range_s: [number, number];
  hp_recommended: boolean;
  precision: string;
  n_nuclides: number;
}

export interface EvaluateOk {
  axis: DoseAxis;
  unit: string;
  times_s: number[];
  nuclides: string[];
  series: Record<string, number[]>;
  peak_atoms: number;
  floor_atoms: number;
}

/** One DAG node = one closure nuclide (or the SF pseudo-sink). Shape mirrors
 *  `engine/chain.build_dag`. `Z/A/N` are null only for the SF sink; `half_life_s`
 *  is null for stable nodes (and the sink). Per-emission energies are NOT here —
 *  they live in the emissions dataset, surfaced in the M6f-2 per-line dose table. */
export interface ChainNode {
  id: string;
  /** Present only on the SF pseudo-sink ("spontaneous fission products"). */
  label?: string;
  Z: number | null;
  A: number | null;
  N: number | null;
  /** Metastable state suffix ("m"/"n") or "" for the ground state. */
  state: string;
  half_life_s: number | null;
  half_life_readable: string;
  stable: boolean;
  decay_modes: string[];
}
export interface ChainEdge {
  source: string;
  target: string;
  /** Decay mode label (α, β⁻, EC, IT, SF, …). */
  mode: string;
  /** Branching fraction in [0, 1]. */
  branching: number;
}
export interface ChainOk {
  nodes: ChainNode[];
  edges: ChainEdge[];
}

export interface DoseWarning {
  reason?: string;
  nuclide?: string;
  [k: string]: unknown;
}

export interface DoseOk {
  quantity: string;
  si_unit: string;
  per: string;
  times_s: number[];
  distance_m: number;
  rate_si: number[];
  scoring_floor_MeV: number;
  warnings: DoseWarning[];
  /** β skin-dose responses only (M6g): the secondary bremsstrahlung γ-dose series scored
   *  when a shield stops the β ("more lead → more photon dose"). null when no shield. A
   *  γ (Sv/air-kerma) quantity — shown beside, NEVER summed into, the β skin (Gy) number. */
  bremsstrahlung?: DoseOk | null;
  /** neutron_dose responses only (M7d): the SOURCE-correlated reaction γ-dose series (e.g.
   *  AmBe 4.438 MeV, scored through the γ engine in the SAME Sv quantity) — null for Cf-252
   *  (prompt-fission γ continuum unmodeled, §11). A γ contribution that DOES stack into the
   *  Sv total, but is kept distinct from the inventory's decay-γ lines (it is a reaction γ,
   *  not a decay line). */
  source_gamma?: DoseOk | null;
  /** spent_fuel_neutron_dose responses only (M9): per-time fraction of the SF neutron source
   *  from emitters without an evaluated ν̄ (chiefly Cm-246 at long cooling) — NOT in the dose.
   *  The honest lower-bound gap, surfaced for the UI to show (the dangerous under-count). */
  dropped_sf_frac?: number[];
}

/** One shield material for the M6g picker. `has_buildup` is the γ-shield gate (a material
 *  without ANS-6.4.3 buildup raises in the γ engine; the UI filters the γ picker to it). */
export interface MaterialInfo {
  id: string;
  has_buildup: boolean;
  density_g_cm3: number;
}

export interface MaterialsOk {
  materials: MaterialInfo[];
}

/** The §9 dose-vs-thickness sweep (M6g, Design-A): distance/time-free per-nuclide γ
 *  coefficients `C_n(x)` over `thicknesses_cm`. `x=0` is the exact unshielded baseline.
 *  The client folds `1/4πd²` and the parent activity at the cursor (zero re-fetch). */
export interface DoseThicknessOk {
  quantity: string;
  si_unit: string;
  material: string;
  thicknesses_cm: number[];
  coeff_by_nuclide: Record<string, number[]>;
  warnings: DoseWarning[];
  scoring_floor_MeV: number;
}

/** One scored photon line for the §9 per-line γ table (M6f-2). `coeff_si` is the
 *  DISTANCE- and TIME-free per-decay SI constant (Sv·m²/decay for H*(10)/effective);
 *  the client applies `1/4πd²` and the parent's activity at the cursor. */
export interface DoseLineRow {
  nuclide: string;
  E_MeV: number;
  yield: number;
  origin: string | null;
  coeff_si: number;
}

export interface DoseLinesOk {
  quantity: string;
  si_unit: string;
  lines: DoseLineRow[];
  warnings: DoseWarning[];
  scoring_floor_MeV: number;
}

export interface ReleaseOk {
  handle: string;
  existed: boolean;
}

export interface NuclidesOk {
  /** Every nuclide the engine can solve (rd dataset), natural-sorted, names only. */
  nuclides: string[];
}

export interface RegistrySizeOk {
  /** Number of live solved inventories (handle-leak canary for invariant #2). */
  size: number;
}

/** One §8 spent-fuel catalog source — inventory sourced from validated `data/spent_fuel`. */
export interface SpentFuelSource {
  id: string;
  label: string;
  category: string;
  blurb: string;
  caveat: string | null;
  referenceTimeS: number;
  burnup_GWd_tHM: number;
  enrichment_pct: number;
  /** M9: this vector carries an intrinsic multi-parent SF neutron source (arms the neutron view). */
  hasNeutron: boolean;
  entries: SolveEntry[];
}
export interface SpentFuelCatalogOk {
  sources: SpentFuelSource[];
}

/** One §8 / §13 #5 fallout catalog source — inventory from validated `data/fallout` (ENDF
 *  U-235 cumulative fission yields). Loaded with `unit="atoms"` at t=0 ≈ H+1 h. */
export interface FalloutSource {
  id: string;
  label: string;
  category: string;
  blurb: string;
  caveat: string | null;
  referenceTimeS: number;
  entries: SolveEntry[];
}
export interface FalloutCatalogOk {
  sources: FalloutSource[];
}

/** Decay-heat (thermal power, W) series over a time grid (M7c §5). */
export interface DecayHeatOk {
  quantity: "decay_heat";
  si_unit: "W";
  times_s: number[];
  total_W: number[];
  by_nuclide_W: Record<string, number[]>;
  coeff_W_per_Bq: Record<string, number>;
  E_rec_MeV: Record<string, number>;
  definition: string;
}

export type SolveResponse = Result<SolveOk>;
export type NuclidesResponse = Result<NuclidesOk>;
export type MaterialsResponse = Result<MaterialsOk>;
export type DoseThicknessResponse = Result<DoseThicknessOk>;
export type RegistrySizeResponse = Result<RegistrySizeOk>;
export type EvaluateResponse = Result<EvaluateOk>;
export type ChainResponse = Result<ChainOk>;
export type DoseResponse = Result<DoseOk>;
export type DoseLinesResponse = Result<DoseLinesOk>;
export type ReleaseResponse = Result<ReleaseOk>;
export type SpentFuelCatalogResponse = Result<SpentFuelCatalogOk>;
export type FalloutCatalogResponse = Result<FalloutCatalogOk>;
export type DecayHeatResponse = Result<DecayHeatOk>;

// --- the client --------------------------------------------------------------

/** The raw Pyodide module proxy: each member is a `(str, ...) -> str` callable. */
type BridgeModule = Record<string, (...args: string[]) => string>;

export class BridgeClient {
  private readonly mod: BridgeModule;

  constructor(mod: BridgeModule) {
    this.mod = mod;
  }

  private call<T>(fn: string, ...args: string[]): T {
    const raw = this.mod[fn];
    if (typeof raw !== "function") {
      throw new Error(`engine.bridge has no callable ${JSON.stringify(fn)}`);
    }
    const out = raw(...args);
    if (typeof out !== "string") {
      throw new Error(`engine.bridge.${fn} returned ${typeof out}, expected JSON text`);
    }
    return JSON.parse(out) as T;
  }

  /** The add-by-name source for the inventory panel; one fetch, cached by the caller. */
  nuclides(): NuclidesResponse {
    return this.call<NuclidesResponse>("nuclides");
  }

  /** The M6g shield-builder material list (id, has_buildup, density); one fetch, cached. */
  materials(): MaterialsResponse {
    return this.call<MaterialsResponse>("materials");
  }

  solve(spec: SolveSpec): SolveResponse {
    return this.call<SolveResponse>("solve", JSON.stringify(spec));
  }

  evaluate(handle: Handle, req: EvaluateRequest): EvaluateResponse {
    return this.call<EvaluateResponse>("evaluate", handle, JSON.stringify(req));
  }

  chain(handle: Handle): ChainResponse {
    return this.call<ChainResponse>("chain", handle);
  }

  dose(handle: Handle, req: DoseRequest): DoseResponse {
    return this.call<DoseResponse>("dose", handle, JSON.stringify(req));
  }

  /** Per-line γ decomposition for the §9 per-line table (M6f-2). Distance/time-free:
   *  the client folds in `1/4πd²` and the cursor activity (no re-fetch on scrub). */
  dose_lines(handle: Handle, req: { quantity?: string; geometry?: string | null; shield?: ShieldSpec | null; medium?: string }): DoseLinesResponse {
    return this.call<DoseLinesResponse>("dose_lines", handle, JSON.stringify(req));
  }

  /** The §9 dose-vs-thickness sweep (M6g): per-nuclide γ coefficients over a thickness grid.
   *  Distance/time-free — the client folds `1/4πd²` + cursor activity (no re-fetch on scrub). */
  dose_thickness(handle: Handle, req: DoseThicknessRequest): DoseThicknessResponse {
    return this.call<DoseThicknessResponse>("dose_thickness", handle, JSON.stringify(req));
  }

  beta_dose(handle: Handle, req: Record<string, unknown>): DoseResponse {
    return this.call<DoseResponse>("beta_dose", handle, JSON.stringify(req));
  }

  neutron_dose(handle: Handle, req: Record<string, unknown>): DoseResponse {
    return this.call<DoseResponse>("neutron_dose", handle, JSON.stringify(req));
  }

  /** The §6.3 spent-fuel SF neutron dose (M9) — multi-parent (S(t)=Σ yield_n·A_n(t)); same
   *  DoseOk shape as `neutron_dose`. `req`: {times_s, source_id, quantity, distance_m, geometry}. */
  spent_fuel_neutron_dose(handle: Handle, req: Record<string, unknown>): DoseResponse {
    return this.call<DoseResponse>("spent_fuel_neutron_dose", handle, JSON.stringify(req));
  }

  /** The §8 spent-fuel catalog (inventory from validated `data/spent_fuel`); one fetch,
   *  merged into the source picker by the store. */
  spent_fuel_catalog(): SpentFuelCatalogResponse {
    return this.call<SpentFuelCatalogResponse>("spent_fuel_catalog");
  }

  /** The §8 / §13 #5 fallout catalog (inventory from validated `data/fallout`); one fetch,
   *  merged into the source picker by the store. */
  fallout_catalog(): FalloutCatalogResponse {
    return this.call<FalloutCatalogResponse>("fallout_catalog");
  }

  /** Decay-heat (W) series over a time grid (M7c §5). One evaluate per (inventory ×
   *  source-age); distance/quantity-free (heat is total locally-deposited power). */
  decay_heat(handle: Handle, req: { times_s: number[] }): DecayHeatResponse {
    return this.call<DecayHeatResponse>("decay_heat", handle, JSON.stringify(req));
  }

  release(handle: Handle): ReleaseResponse {
    return this.call<ReleaseResponse>("release", handle);
  }

  /** Live-handle count — the leak canary the gate asserts is 1 across re-solves. */
  registry_size(): RegistrySizeResponse {
    return this.call<RegistrySizeResponse>("registry_size");
  }
}
